### ISSUE TITLE
Order of attributes in to_xml output is unstable

### ISSUE THREAD

**Issue by mickdekkers** [ballsteve/xrust#98]
Hi,

I've just set up [snapshot testing](https://github.com/mitsuhiko/insta) in my project that uses xrust, but unfortunately it would seem that `to_xml` is currently unsuitable for this purpose, because the serialized order of attributes is not consistent between different `parse` + `to_xml` calls:

![image](https://github.com/user-attachments/assets/2813a460-42a1-4982-961b-066ed8e76063)

I've set up a repo with a minimal reproducible example: https://github.com/mickdekkers/xrust-serialization-attribute-order-instability-issue-repro

Ideally, the order of attributes in the source XML would be maintained unless the `Node`s are explicitly mutated to change this order.

The issue occurs with the code on `main` as well as `dev`. The repro points to `dev`.

Would once again appreciate it if you could take a look! 😄 

Many thanks!

**Comment by ballsteve** [comment]
It sounds like you're after a Canonical XML output. Daniel has added in some support for that, but I think a good design idea would be to make it an option in the OutputDefinition; that way you would use to_xml_with_options(), just like indenting.

**Comment by mickdekkers** [comment]
Hi @ballsteve, thanks for the suggestion! Canonical XML does seem to cover most<sup>1</sup> of what I'm after 😄 

However, I just tried replacing [line 13 of `main.rs`](https://github.com/mickdekkers/xrust-serialization-attribute-order-instability-issue-repro/blob/2ef0bc6ab52d5e9a2103ec38b33c63be19c49a5c/src/main.rs#L13) to get Canonical XML output:
```diff
- let xml_output = doc.to_xml();
+ let xml_output = doc.get_canonical().unwrap().to_xml();
```
and the attribute order of the serialized XML string is unfortunately still changing:

![image](https://github.com/user-attachments/assets/41832462-8fb2-417a-bf54-89b23ab6ab61)

another run output:

![image](https://github.com/user-attachments/assets/7874faca-99ba-4f1d-91f3-56fb1fb8bf47)

another:

![image](https://github.com/user-attachments/assets/b2ee5e85-a4be-45ec-91ed-a90a315c41a6)

Do you or @Devasta perhaps know what might be causing this? As far as I can tell based on a few dozen runs, it's only the _order of attributes_ that's unstable, but not e.g. the order of element nodes within a parent node.

<sup>1</sup> I'm working on an [RFC 5261](https://datatracker.ietf.org/doc/html/rfc5261) implementation as one of the libraries used in my project. Implementing full compliance around e.g. whitespace handling on [`<remove>`](https://datatracker.ietf.org/doc/html/rfc5261#section-4.5) would be nice, but is a non-goal for now

**Comment by ballsteve** [comment]
I don't know what's causing that. There are two issues to resolve here:

1. Make sure that the canonical output conforms to the W3C spec, and
2. Change the design to include canonical output in the OutputDefinition struct.

Both of those are on the development list. We'll get to them ASAP.

**Comment by Devasta** [comment]
So I think the issue is with the storage of attributes in Smite.

https://github.com/ballsteve/xrust/blob/2449e4dfe85d613d8d096fed21f7c1e2f768e0f9/src/trees/smite.rs#L73

Hashmaps in Rust do not preserve order.

XML does not care about attribute order, however the [XDM Spec](https://www.w3.org/TR/xpath-datamodel/#dm-attributes) says "The order of Attribute Nodes is stable but implementation dependent."

I'll look into using a BTree instead, see if there are any issues.

@mickdekkers You only need the output to be stable, right? I don't think we can or want to care about attribute order on the input document.

**Comment by Devasta** [comment]
https://github.com/ballsteve/xrust/tree/dev_issue_98

Made a change here, let me know if this is sufficient?

**Comment by Devasta** [comment]
Good point, I'll take a look at namespace nodes tonight and see what needs doing there.

Also very interested to hear about your project implementing RFC5261, do share it with us if you can. 😄

**Comment by Devasta** [comment]
Right, switching the namespace nodes to a BTreeMap from Hashmap is trivial, however I have found another issue with serialization that I have opened as a separate issue https://github.com/ballsteve/xrust/issues/99

Will try to fix that, then I will push both to our dev branch.

**Comment by ballsteve** [comment]
This issue should be resolved in release 1.1, so closing this issue.
